### PR TITLE
Serverinfo refactoring

### PR DESCRIPTION
**Pull request by bmorel**
No behavior change nor bugfix are expected, this only adds documentation, and improves code readability by using RAII. Need that work for future planned features.

### CODE REVIEW

**Comment by MoonPadUSer** [comment]
It looks beautiful

File: src/engine/serverbrowser.cpp [hunk 1]
```diff
@@ -273,32 +253,35 @@ void pingservers()
...
     uchar ping[MAXTRANS];
+    // ENet is crappy enough to invert it's members depending on
+    // target OS.
+    // This does also imply that this "efficient" library wastes 25%
+    // bytes in alignment on windows for this structure.
+#ifdef WIN32
+    ENetBuffer buf = { 0, ping };
+#else
+    ENetBuffer buf = { ping, 0 };
+#endif
```
Comment (robalni) [inline src/engine/serverbrowser.cpp:266]: Why not do it like this, like we do in other places in the code?

```
    ENetBuffer buf;
    buf.data = ping;
    buf.dataLength = 0;
```

Then we don't need to care about the order.

Comment (TheAssassin) [reply src/engine/serverbrowser.cpp:266]:
```cpp
    ENetBuffer buf{};
```

Otherwise, the struct isn't initialized properly. But I'd just

```cpp
    ENetBuffer buf{0, ping};
```

File: src/shared/defs.hpp [hunk 2]
```diff
@@ -0,0 +1,10 @@
+#ifndef DEFS_HPP
+#define DEFS_HPP
+
+#define SERVER_PORT 28801
+#define MAXSDESCLEN 80
+#define MAXBRANCHLEN 16
+#define MAXTRANS 5000   // max amount of data to swallow in 1 go
+#define VERSION_GAME 230
+
+#endif
```
Comment (robalni) [inline src/shared/defs.hpp:10]: For this file and also serverinfo.hpp: We use .h as file extension for all other header files. Maybe we should keep it the same (until we change all of them if we ever do that).

File: src/engine/serverbrowser.cpp (hunk 1 above)
Comment (robalni) [reply src/engine/serverbrowser.cpp:266]: There are two problems with `ENetBuffer buf{0, ping};`:

1. We need to write different code for different operating systems as seen in this code.
2. We can't see that the right members get the right values; we have to rely on knowing the order. I like the designated initializers in C99 but if we don't have them I think the best we can do, if we want to make sure everything is initialized, is:

```
    ENetBuffer buf{};
    buf.data = ping;
    buf.dataLength = 0;
```

Comment (TheAssassin) [reply src/engine/serverbrowser.cpp:266]: Well, I'm not a big fan of relying on the order either, and even if it's rather unlike that things might go wrong with regards to that, it's also important to keep things consistent in the code. Some day, we might make ENet a more usable C++ library, but until this day has come, let's just keep things consistent.

We should really turn on this warning in the compiler that complains about uninitialized structs...

File: src/engine/serverbrowser.cpp [hunk 3]
```diff
@@ -309,12 +292,9 @@ void checkresolver()
     int resolving = 0;
     loopv(servers)
     {
-        serverinfo &si = *servers[i];
-        if(si.resolved == serverinfo::RESOLVED) continue;
-        if(si.address.host == ENET_HOST_ANY)
+        if( servers[i]->need_resolve( resolving ) )
```
Comment (TheAssassin) [inline src/engine/serverbrowser.cpp:295]: From the method name and its param, I could not guess what exactly that method does. You could get away by incrementing the `resolving` variable inside this method. That's what was done before, too. I checked the implementation, and I consider it bad practice the way it's done now.

File: src/shared/serverinfo.cpp [hunk 4]
```diff
@@ -0,0 +1,507 @@
...
+}
+
+bool serverinfo::need_resolve( int& resolving )
+{
+    if( m_resolved == RESOLVED || m_address.host != ENET_HOST_ANY )
+    {
+        return false;
+    }
+    int ret = m_resolved == UNRESOLVED;
+    ++resolving;
```
Comment (TheAssassin) [inline src/shared/serverinfo.cpp:413]: That shouldn't be done inside this method. The method does not make any use of the variable.

File: src/engine/serverbrowser.cpp [hunk 5]
```diff
@@ -309,12 +292,9 @@ void checkresolver()
...
         {
-            if(si.resolved == serverinfo::UNRESOLVED) { si.resolved = serverinfo::RESOLVING; resolverquery(si.name); }
-            resolving++;
+            resolverquery(servers[i]->name());
```
Comment (TheAssassin) [inline src/engine/serverbrowser.cpp:297]: Just increment `resolving` here.

File: src/engine/serverbrowser.cpp [hunk 6]
```diff
@@ -309,12 +292,9 @@ void checkresolver()
...
         }
     }
     if(!resolving) return;
```
Comment (TheAssassin) [inline src/engine/serverbrowser.cpp:300]: Please replace by some explicit `> 0` check to increase readability.

File: src/engine/serverbrowser.cpp [hunk 7]
```diff
@@ -326,18 +306,15 @@ void checkresolver()
...
-                si.resolved = serverinfo::RESOLVED;
-                si.address.host = addr.host;
                 break;
             }
         }
     }
 }
 
-static int lastreset = 0;
+int lastreset = 0;
```
Comment (TheAssassin) [inline src/engine/serverbrowser.cpp:317]: What's the motivation behind this change? Just curious.

File: src/engine/serverbrowser.cpp [hunk 8]
```diff
@@ -218,35 +220,13 @@ int connectwithtimeout(ENetSocket sock, const char *hostname, const ENetAddress
...
-    sortedservers = false;
-
-    return si;
-}
-
 void addserver(const char *name, int port, int priority, const char *desc, const char *handle, const char *flags, const char *branch)
 {
-    loopv(servers) if(!strcmp(servers[i]->name, name) && servers[i]->port == port) return;
-    if(newserver(name, port, priority, desc, handle, flags, branch) && verbose >= 2)
+    loopv(servers) if( servers[i]->is_same( name, port ) ) return;
```
Comment (TheAssassin) [inline src/engine/serverbrowser.cpp:228]: I'm not a big fan of this kind of method. Maybe it's also just the naming that isn't entirely clear.

File: src/engine/serverbrowser.cpp [hunk 9]
```diff
@@ -346,63 +323,34 @@ void checkpings()
...
-            getstring(text, p);
-            si->players.add(newstring(text));
-        }
-        if(gver >= 225) loopi(si->numplayers)
+        loopv(servers)
         {
-            if(p.overread()) break;
-            getstring(text, p);
-            si->handles.add(newstring(text));
+            if( servers[i]->is_same( addr ) )
```
Comment (TheAssassin) [inline src/engine/serverbrowser.cpp:335]: I'm really not a fan of this method. It is not a full equality check, only checks whether the address is the same. This should be reflected by the method name, too. As a reader, I initially got the impression that it advertises more than what it actually provides.

A more elegant way would be to implement a proper operator for the address. This could even be done without modifying whatever component the `addr` object is coming from (likely ENet).

File: src/engine/serverbrowser.cpp [hunk 10]
```diff
@@ -346,63 +323,34 @@ void checkpings()
...
+            {
+                si = servers[i];
+                break;
+            }
         }
-        sortedservers = false;
+        if(!si && searchlan) si = serverinfo::newserver(NULL, addr.port-1, 1, NULL, NULL, NULL, NULL, addr.host);
```
Comment (TheAssassin) [inline src/engine/serverbrowser.cpp:341]: TODO for self: check what `newserver` does (likely some sort of factory).

Comment (TheAssassin) [reply src/engine/serverbrowser.cpp:341]: Please use explicit pointer checks, especially in new code. That helps increase the readability.

File: src/engine/serverbrowser.cpp [hunk 11]
```diff
@@ -346,63 +323,34 @@ void checkpings()
...
+        if(!si) continue;
```
Comment (TheAssassin) [inline src/engine/serverbrowser.cpp:342]: Please use explicit pointer checks, especially in new code. That helps increase the readability.

File: src/engine/serverbrowser.cpp [hunk 12]
```diff
@@ -346,63 +323,34 @@ void checkpings()
...
+        si->update( buf.dataLength, buf.data, serverdecay, totalmillis, lastreset );
```
Comment (TheAssassin) [inline src/engine/serverbrowser.cpp:343]: This `update` method is like a bundled setter, and as far as I can see only "saves" a few lines. TODO for self: check if it does anything other than setting those attributes.

File: src/engine/serverbrowser.cpp [hunk 13]
```diff
@@ -526,8 +470,32 @@ void writeservercfg()
     f->printf("// servers which are connected to or queried get added here automatically\n\n");
     loopv(servers)
     {
-        serverinfo *s = servers[i];
-        f->printf("addserver %s %d %d %s %s %s %s\n", s->name, s->port, s->priority, escapestring(s->sdesc[0] ? s->sdesc : s->name), escapestring(s->authhandle), escapestring(s->flags), escapestring(s->branch));
+        servers[i]->writecfg( *f );
```
Comment (TheAssassin) [inline src/engine/serverbrowser.cpp:473]: Minor naming issue (this should be `write_config` or `write_cfg`, like the remaining methods).

File: src/engine/serverbrowser.cpp [hunk 14]
```diff
@@ -526,8 +470,32 @@ void writeservercfg()
...
     }
     delete f;
 }
+
+VAR(IDF_PERSIST, hideincompatibleservers, 0, 0, 1);
+void getservers(int server, int prop, int idx)
```
Comment (TheAssassin) [inline src/engine/serverbrowser.cpp:479]: Blank line above would be appreciated.

File: src/engine/serverbrowser.cpp [hunk 15]
```diff
@@ -526,8 +470,32 @@ void writeservercfg()
...
+    if(server < 0)
+    {
+        // this codes does not really hide stuff, it _removes_ stuff
+        if (hideincompatibleservers)
+        {
+            //servers.erase( std::remove_if(), servers.end()); refuses to work here, dunno why
+            //so for now let's keep the waste
+            vector<serverinfo *> servers_new;
+            std::copy_if( servers.begin(), servers.end(),
+                std::back_inserter( servers_new ),serverinfo::server_compatible );
```
Comment (TheAssassin) [inline src/engine/serverbrowser.cpp:490]:
```suggestion
                std::back_inserter( servers_new ), serverinfo::server_compatible );
```

File: src/engine/serverbrowser.cpp [hunk 16]
```diff
@@ -526,8 +470,32 @@ void writeservercfg()
...
+            servers = servers_new;
+        }
+
+        intret(servers.length());
+    }
+    else if(servers.inrange(server))
+    {
+        servers[server]->cube_get_property( prop, idx );
```
Comment (TheAssassin) [inline src/engine/serverbrowser.cpp:498]: Any reason this method name is prefixed with `cube_`? TODO for myself: check corresponding header.

File: src/game/client.cpp [hunk 17]
```diff
@@ -3240,142 +3240,29 @@ namespace client
         }
     }
 
-    int serverstat(serverinfo *a)
+    bool serverinfocompare(serverinfo const*a, serverinfo const*b)
```
Comment (TheAssassin) [inline src/game/client.cpp:3243]: New naming convention would suggest `serverinfo_compare`. But I do wonder why this is not implemented as an operator directly?

Comment (TheAssassin) [reply src/game/client.cpp:3243]: Also, missing spaces.

```suggestion
    bool serverinfocompare(serverinfo const* a, serverinfo const* b)
```

File: src/game/client.cpp [hunk 18]
```diff
@@ -3240,142 +3240,29 @@ namespace client
...
     {
-        if(a->attr.length() > 4 && a->numplayers >= a->attr[4])
+        int comp = a->version_compare( *b );
+        if( comp != 0 )
```
Comment (TheAssassin) [inline src/game/client.cpp:3246]: The motivation behind these calls is not clear. There should be additional documentary comments in the code.

File: src/game/client.cpp [hunk 19]
```diff
@@ -3240,142 +3240,29 @@ namespace client
...
-                    retsw(aa->priority, ab->priority, false);
-                    break;
-                }
-                default: break;
-            }
+            comp = a->compare(*b, style, reverse);
+            if( comp != 0 ) return comp < 0;
         }
-        return strcmp(a->name, b->name);
+        return strcmp( a->name(), b->name() ) < 0;
```
Comment (TheAssassin) [inline src/game/client.cpp:3265]: Meaning changed in comparison to old code (should likely be `!= 0`, right?). If this is *really* intentional, then there should be a comment explaining why you consider all values `>= 0` here.

File: src/shared/defs.hpp (hunk 2 above)
Comment (TheAssassin) [reply src/shared/defs.hpp:10]: Also, this change is not strictly related to the topic of the PR, as far as I can see (it also missed a few other definitions).

Can someone please create an issue so we can move all remaining definitions into this file properly, later on? Please don't forget to include a link to this comment.

File: src/shared/serverinfo.hpp [hunk 20]
```diff
@@ -0,0 +1,231 @@
...
+#include <string>
+
+#include "defs.hpp"
+
+struct serverinfo;
+
+// list of known servers, with or without informations
+// TODO:
+//   * should be a private static of serverinfo
+//   * stop using raw pointers (RAII ftw)
```
Comment (TheAssassin) [inline src/shared/serverinfo.hpp:14]: Good idea. This is the ideal use case for shared pointers.

File: src/shared/serverinfo.hpp [hunk 21]
```diff
@@ -0,0 +1,231 @@
...
+#include <vector>
+#include <string>
+
+#include "defs.hpp"
+
+struct serverinfo;
+
+// list of known servers, with or without informations
+// TODO:
+//   * should be a private static of serverinfo
```
Comment (TheAssassin) [inline src/shared/serverinfo.hpp:13]: Well, ideally it should not be static at all. The goal should be to no longer depend on static memory for the game's state.

File: src/shared/serverinfo.hpp [hunk 22]
```diff
@@ -0,0 +1,231 @@
...
+//   * stop using raw pointers (RAII ftw)
+extern vector<serverinfo*> servers;
+
+// code imported, here to reduce build dependencies
```
Comment (TheAssassin) [inline src/shared/serverinfo.hpp:17]: These are not "imports", they're forward declarations. Is that what you meant? Or do I miss something here?

There should be a TODO to remove those again before merging. They might've helped speed up your development, however they increase coupling between code entities a lot.

File: src/shared/serverinfo.hpp [hunk 23]
```diff
@@ -0,0 +1,231 @@
...
+extern bool filterstring(char *dst, const char *src, bool newline, bool colour, bool whitespace, bool wsstrip, size_t len);
+namespace client
```
Comment (TheAssassin) [inline src/shared/serverinfo.hpp:19]: There should be a blank line above this line, like you have one below this block.

File: src/shared/serverinfo.hpp [hunk 24]
```diff
@@ -0,0 +1,231 @@
...
+}
+
+//this enum lists (some of) the server properties that can be used for sorting.
+//The C and C++ specs says and enum starts with the value 0 except if explicitly
+//specified.
+//It have 2 list of members:
+// * row 1: original names, for compat, and maybe to allow some naming support in
+//          CS latter
+// * row 2: aliases to avoid reapeating 'SINFO::' for nothing.
+// TODO: use scoped enums
```
Comment (TheAssassin) [inline src/shared/serverinfo.hpp:36]: Grammar is a bit off. Also, I don't think we need to explain C++ standards there.

```suggestion
/**
 * (some of) the server properties that can be used for sorting
 *
 * it has 2 types of members:
 *  - row 1: original names, for compat, and maybe to allow some naming support in CS later
// * row 2: aliases to avoid reapeating 'SINFO::' for nothing.
// TODO: use scoped enums
```

Even though all enums are initialized with `0`, I tend to hardcode the starting value in enums by the way, following the logic "it's not a good idea if you have to explain it". I think that's also done in the remaining codebase.

File: src/shared/serverinfo.hpp [hunk 25]
```diff
@@ -0,0 +1,231 @@
...
+    MM_LOCKED  ,
+    MM_PRIVATE ,
+    MM_PASSWORD,
+};
+
+struct serverinfo
+{
+    // should probably be private, but:
+    // WAITING is used in serverbrowser for cubescript exposition
+    enum
```
Comment (TheAssassin) [inline src/shared/serverinfo.hpp:83]: Would it make sense to make those two values static const(expr) variables? They don't seem to have to do a lot with each other...

File: src/shared/serverinfo.hpp [hunk 26]
```diff
@@ -0,0 +1,231 @@
...
+    {
+        MAXPINGS = 3,
+        WAITING = INT_MAX
+    };
+
+private:
+    // list of player/spectators names
+    std::vector<std::string> m_players;
```
Comment (TheAssassin) [inline src/shared/serverinfo.hpp:91]: TODO: discuss whether to use any form of prefix. There's a lot of pros and cons against them. I've used all kinds of prefixes (and also omitted them entirely), and I have no clear preference. In my own upcoming PR I used a simple trailing underscore...

File: src/shared/serverinfo.hpp [hunk 27]
```diff
@@ -0,0 +1,231 @@
...
+    // name resolution state.
+    enum
+    {
+        UNRESOLVED,
+        RESOLVING,
+        RESOLVED
+    } m_resolved;
+
+    // is "servers" sorted?
+    static bool s_sortedservers;
```
Comment (TheAssassin) [inline src/shared/serverinfo.hpp:131]: Could we dump that? We can just sort before every use, it shouldn't be a very expensive operation. Keeping this kind of state appears somewhat dangerous to me.

File: src/shared/serverinfo.hpp [hunk 28]
```diff
@@ -0,0 +1,231 @@
...
+    //
+    // Styles NUMPLRS and PRIO are "better if higher".
+    // Others are "better if lower".
+    // Reverse invert the sorting.
+    //
+    // returns:
+    //   * -1 if *this < other
+    //   *  0 if *this = other
+    //   *  1 if *this > other
+    int compare( serverinfo const& other, int style, bool reverse ) const;
```
Comment (TheAssassin) [inline src/shared/serverinfo.hpp:157]: Do we absolutely need both overloads? I think one should just cast before use if needed.

File: src/shared/serverinfo.hpp [hunk 29]
```diff
@@ -0,0 +1,231 @@
...
+    int compare( serverinfo const& other, SINFO style, bool reverse ) const;
+
+    // return 0 if versions match, -1 if *this > other and 1 otherwise
+    int version_compare( serverinfo const& other ) const;
```
Comment (TheAssassin) [inline src/shared/serverinfo.hpp:161]: Method naming is pretty inconsistent. Should likely be called `compare_version`.

File: src/shared/serverinfo.hpp [hunk 30]
```diff
@@ -0,0 +1,231 @@
...
+    // "return" a property of this server to CS
+    // if property < 0: returns the number of properties one can query (4)
+    // if property > 3: do nothing.
+    // if property > 0:
+    //   if index < 0: returns the size of requested array
+    //   if index > array.size(): do nothing
+    //   if property == 1: returns m_attr[index] (integer)
+    //   if property == 2: returns players[index] (result?)
+    //   if property == 3: returns handles[index] (result?)
+    // if propert == 1:
```
Comment (TheAssassin) [inline src/shared/serverinfo.hpp:178]:
```suggestion
    // if property == 1:
```

File: src/shared/serverinfo.hpp [hunk 31]
```diff
@@ -0,0 +1,231 @@
...
+    //   if index ==  3: returns description or name if no description (result)
+    //   if index ==  4: returns map played (result)
+    //   if index ==  5: returns number of players (integer)
+    //   if index ==  6: returns ping? Guess it's roundtrip time? (integer)
+    //   if index ==  7: returns m_lastinfo? guess it's time of last update? (integer)
+    //   if index ==  8: returns authhandle?? (result)
+    //   if index ==  9: returns flags??  (result)
+    //   if index == 10: returns branch?? (resull)
+    //   if index == 11: returns priority (integer)
+    void cube_get_property( int property, int index );
```
Comment (TheAssassin) [inline src/shared/serverinfo.hpp:191]: What does `cube` refer to? The docstring implies this method is somewhat useful with regards to CubeScript. `cube` != `CubeScript`, `cube` is a game whereas CubeScript is a language. I find it a bit confusing.

Other than that, there should be a TODO to invite further contributors into reverse engineering the other values and answer the open questions.

File: src/shared/serverinfo.hpp [hunk 32]
```diff
@@ -0,0 +1,231 @@
...
+
+    // write some current server's attributes into file
+    // * m_name
+    // * port
+    // * priority
+    // * description() (escaped)
+    // * authhandle (escaped)
+    // * flags (escaped)
+    // * branch (escaped)
+    void writecfg( stream& file ) const;
```
Comment (TheAssassin) [inline src/shared/serverinfo.hpp:201]: Naming is a bit off, see my other comment. But it's certainly a good idea to have such a method.

*(With regards to the SRP, there should be a separate class for this kind of output, though. No action required, but it should be said once.)*

File: src/shared/serverinfo.hpp [hunk 33]
```diff
@@ -0,0 +1,231 @@
...
+
+    // update server's internal informations.
+    // According to what exactly is still unknown to me.
+    // Candidate for rewrite.
+    // TODO: make private?
+    //   * only used in serverbrowser.cpp
+    void update( size_t len, void const* data, int serverdecay, int totalmillis, int lastreset );
```
Comment (TheAssassin) [inline src/shared/serverinfo.hpp:208]: Should probably be provided through setters only. To the outside user, this just looks like it collapses 5 lines into one. See my other comment, too.

File: src/shared/serverinfo.hpp [hunk 34]
```diff
@@ -0,0 +1,231 @@
...
+
+    // methods that allows multi-threaded server name resolution without breaking encapsulation
+    bool validate_resolve( char const* name, ENetAddress const& addr );
```
Comment (TheAssassin) [inline src/shared/serverinfo.hpp:211]: There need to be proper docstrings. To me as a reader, it doesn't get clear what would be "validated" here.

File: src/shared/serverinfo.hpp [hunk 35]
```diff
@@ -0,0 +1,231 @@
...
+    bool need_resolve( int& resolving );
+
+    // check if current's server network info (IP/port) match parameter(s)
+    bool is_same( ENetAddress const& addr ) const;
```
Comment (TheAssassin) [inline src/shared/serverinfo.hpp:215]: As written elsewhere, the first prototype should not exist. You track the address as an instance of `ENetAddress` anyway. The more elegant approach is IMO to just implement an `operator==(const ENetAddress&, const ENetAddress&)`, and provide a getter here instead.

File: src/shared/serverinfo.hpp [hunk 36]
```diff
@@ -0,0 +1,231 @@
...
+    bool is_same( char const* oname, int oport ) const;
```
Comment (TheAssassin) [inline src/shared/serverinfo.hpp:216]: Not a fan of this method. I think it should be removed from the class, or be named differently to avoid giving users a wrong impression of what the method does. To me as a normal reader, it actually looks more or less like a holistic comparison operator, but it's not.

File: src/shared/serverinfo.hpp [hunk 37]
```diff
@@ -0,0 +1,231 @@
...
+
+    // sends a ping to target
+    void ping( ENetSocket& sock, int serverdecay, int millis );
```
Comment (TheAssassin) [inline src/shared/serverinfo.hpp:219]: Breaks the SRP to some extent.

What are the side effects of this call? This apparently likely triggers something in ENet.

File: src/shared/serverinfo.hpp [hunk 38]
```diff
@@ -0,0 +1,231 @@
...
+
+    // register a new server into "servers"
+    static serverinfo *newserver(const char *name, int port = SERVER_PORT, int priority = 0, const char *desc = nullptr, const char *handle = nullptr, const char *flags = nullptr, const char *branch = nullptr, uint ip = ENET_HOST_ANY);
+
+    // check if the target is compatible with current client
+    static bool server_compatible( serverinfo const* target );
```
Comment (TheAssassin) [inline src/shared/serverinfo.hpp:225]: Naming wise, this should likely be prefixed `is_`. The motivation doesn't get clear from the docstring, though. Could you please explain what "compatible" means?

File: src/shared/serverinfo.hpp [hunk 39]
```diff
@@ -0,0 +1,231 @@
...
+
+    // check if current's server network info (IP/port) match parameter(s)
+    bool is_same( ENetAddress const& addr ) const;
+    bool is_same( char const* oname, int oport ) const;
+
+    // sends a ping to target
+    void ping( ENetSocket& sock, int serverdecay, int millis );
+
+    // register a new server into "servers"
+    static serverinfo *newserver(const char *name, int port = SERVER_PORT, int priority = 0, const char *desc = nullptr, const char *handle = nullptr, const char *flags = nullptr, const char *branch = nullptr, uint ip = ENET_HOST_ANY);
```
Comment (TheAssassin) [inline src/shared/serverinfo.hpp:222]: Is this an implementation of the "named constructor" idiom, where it also inserts the server into the array?

File: src/shared/serverinfo.hpp [hunk 40]
```diff
@@ -0,0 +1,231 @@
...
+
+    // check if the target is compatible with current client
+    static bool server_compatible( serverinfo const* target );
+
+    // sorts "servers"
+    static void sort( vector<serverinfo*> &servers );
```
Comment (TheAssassin) [inline src/shared/serverinfo.hpp:228]: Do we even need the parameter? The `newserver` method also doesn't need it.

File: src/shared/tools.h [hunk 41]
```diff
@@ -672,6 +673,11 @@ template <class T> struct vector
         add();
     }
 
+    void push_back( T const& el )
```
Comment (TheAssassin) [inline src/shared/tools.h:676]: I've missed that method recently in my own upcoming PR, too. In the future, please consider sending separate PRs for those.

File: src/shared/serverinfo.cpp [hunk 42]
```diff
@@ -0,0 +1,507 @@
...
+    cleanup();
+}
+
+void serverinfo::reset()
+{
+    m_lastping = m_lastinfo = -1;
+    s_sortedservers = false;
+}
+
+int serverinfo::compare( serverinfo const& other, int style, bool reverse ) const
```
Comment (robalni) [inline src/shared/serverinfo.cpp:135]: We usually put `const` on the left side when possible. It's confusing to mix the two styles. This is done in some other places too, not only this line.

File: src/shared/serverinfo.cpp [hunk 43]
```diff
@@ -0,0 +1,507 @@
...
+		}
+		bool higher_is_better = style != SINFO_NUMPLRS && style != SINFO_PRIO;
+		if( higher_is_better != reverse ? ac < bc : ac > bc )
+		{
+			return -1;
+		}
+		if( higher_is_better != reverse ? ac > bc : ac < bc )
+		{
+			return  1;
+		}
```
Comment (robalni) [inline src/shared/serverinfo.cpp:214]: Indentation is weird here.

**Comment by robalni** [comment]
I can't build this. I get the error:

```
[ 30%] Building CXX object src/CMakeFiles/blue-nebula_linux.dir/engine/serverbrowser.cpp.o
/home/a/s/bluenebula/src/engine/serverbrowser.cpp: In function ‘void getservers(int, int, int)’:
/home/a/s/bluenebula/src/engine/serverbrowser.cpp:486:18: error: ‘copy_if’ is not a member of ‘std’; did you mean ‘copy’?
  486 |             std::copy_if( servers.begin(), servers.end(),
      |                  ^~~~~~~
      |                  copy
```

and if I remove that code, as already done in master, I get linker errors about undefined references to functions in shared/serverinfo.cpp and when I look I see that this file is not added to the cmake list of source files. I wonder how Travis could build it.

File: src/shared/serverinfo.cpp (hunk 42 above)
Comment (TheAssassin) [reply src/shared/serverinfo.cpp:135]: The reason is a different one. This use of `const` indicates that the method does not alter the state of the object. That means that the compiler permits its execution also for `const` references. You need to do the same for comparison operators as well. There is no way to put it on the left side.

**Comment by robalni** [comment]
I have noticed that the server browser sort order jumps around every 4 seconds when I test it with these changes (if you have the mouse outside of it). It doesn't do that on master.

File: src/shared/serverinfo.cpp (hunk 42 above)
Comment (robalni) [reply src/shared/serverinfo.cpp:135]: I was talking about the `const` inside the argument list, not after the function. We would normally do `const serverinfo &other`, not `serverinfo const& other`

Comment (TheAssassin) [reply src/shared/serverinfo.cpp:135]: I see. Well, in C++ it's nowadays more or less a matter of personal preference. In good old C times, the order did make a huge difference. Simple example:

```cpp
// one can modify the contents of a
char * const a;
// here, one cannot change the contents, but change the pointer address
const char* b;
```

I'm personally in preference of `const serverinfo& abc`. But we have an issue about code style open anyway. We've agreed on avoiiding such "preference style" discussions until we have a consensus (and ideally a `clang-format` file which can be used to auto-format files to some extent).

Comment (robalni) [reply src/shared/serverinfo.cpp:135]: Yes, order matters if you put two things on the left side of the `const` but not if you put `const` on the left or right side of the leftmost thing.

**Comment by bmorel** [comment]
> I excluded the implementation `serverinfo.cpp` from my review intentionally, but focused on the "outside view". There's a lot of questions, and a few conceptual change requests. Please don't hesitate to answer if you disagree.
> 
> Comments regarding style aren't strictly binding. It's appreciated if you'd apply a few of them, though. Feel free to discuss about them on GitHub as well. That'll help define a proper code style later on.
> 
> By the way (I could not comment those changes directly): Making the `cpp` files no longer executable is a good idea. But why also the shell scripts in `src/scripts`?

The src/scripts? Maybe some are accidents. Detail please, but this one is noisy enough. We should have a whitelist of files with the +x right anyway. And reject PRs that don't follow it without reasons, dunno how this can be done.

File: src/engine/serverbrowser.cpp [hunk 44]
```diff
@@ -273,32 +253,35 @@ void pingservers()
...
+    ENetBuffer buf = { ping, 0 };
+#endif
     ucharbuf p(ping, sizeof(ping));
     putint(p, totalmillis ? totalmillis : 1);
+    buf.dataLength = p.length();
 
     static int lastping = 0;
-    if(lastping >= servers.length()) lastping = 0;
-    loopi(maxservpings ? min(servers.length(), maxservpings) : servers.length())
+    auto num_servers = servers.size();
```
[the remark this answers is not in the document]
Comment (bmorel) [reply src/engine/serverbrowser.cpp:264]: good point for const. For size and opt builds, I prefer to move repetitive calls into a variable. Makes debugging easier since you can inspect the variable, and shorter and thus more readable.

File: src/shared/serverinfo.hpp (hunk 37 above)
Comment (bmorel) [reply src/shared/serverinfo.hpp:221]: it sends a ping.

File: src/shared/tools.h (hunk 41 above)
Comment (bmorel) [reply src/shared/tools.h:676]: originally, it was in a distinct commit, but yes, should have done that.

File: src/engine/serverbrowser.cpp (hunk 6 above)
Comment (bmorel) [reply src/engine/serverbrowser.cpp:292]: you want "if( some_bool_value > 0 )" ?
